File: firebase/functions/lib/cloud_function.dart
import 'package:firebase_functions_interop/firebase_functions_interop.dart';

/// Parent class for various Cloud Functions with multiple types
/// (on_request, on_firestore, and on_call).
///
/// This does not get deployed as a standalone Cloud Function.
abstract class CloudFunction {
  String get functionName;

  void register(FirebaseFunctions functions);
}

File: firebase/functions/src/cloud_function.ts
import { HttpsFunction, CloudFunction as FirebaseCloudFunction } from 'firebase-functions'

/**
 * Base interface for all Cloud Functions.
 */
export interface CloudFunction {
    readonly functionName: string
    register(
        functions: typeof import('firebase-functions'),
        cors: typeof import('cors')
    ): HttpsFunction | FirebaseCloudFunction<unknown> | void
}
